networkx/algorithms/isomorphism/vf2pp_helpers/node_ordering.py

```diff
@@ -0,0 +1,78 @@
+import networkx as nx
+
+
+def _matching_order(graph_params):
+    """The node ordering as introduced in VF2++.
+
+    Notes
+    -----
+    Taking into account the structure of the Graph and the node labeling, the nodes are placed in an order such that,
+    most of the unfruitful/infeasible branches of the search space can be pruned on high levels, significantly
+    decreasing the number of visited states. The premise is that, the algorithm will be able to recognize
+    inconsistencies early, proceeding to go deep into the search tree only if it's needed.
+
+    Parameters
+    ----------
+    graph_params: namedtuple
+        Contains:
+
+            G1,G2: NetworkX Graph or MultiGraph instances.
+                The two graphs to check for isomorphism or monomorphism.
+
+            G1_labels,G2_labels: dict
+                The label of every node in G1 and G2 respectively.
+
+    Returns
+    -------
+    node_order: list
+        The ordering of the nodes.
+    """
+    G1, G2, G1_labels, _, _, nodes_of_G2Labels, _ = graph_params
+    if not G1 and not G2:
+        return {}
+
+    V1_unordered = set(G1.nodes())
+    label_rarity = {label: len(nodes) for label, nodes in nodes_of_G2Labels.items()}
+    used_degrees = {node: 0 for node in G1}
+    node_order = []
+
+    while V1_unordered:
+        rarest_nodes = _all_argmax(
+            V1_unordered, key_function=lambda x: -label_rarity[G1_labels[x]]
+        )
+        max_node = max(rarest_nodes, key=G1.degree)
+
+        for dlevel_nodes in nx.bfs_layers(G1, max_node):
+            nodes_to_add = dlevel_nodes.copy()
+            while nodes_to_add:
+                max_used_deg_nodes = _all_argmax(
+                    nodes_to_add, key_function=lambda x: used_degrees[x]
+                )
+                max_deg_nodes = _all_argmax(
+                    max_used_deg_nodes, key_function=lambda x: G1.degree[x]
+                )
+                next_node = min(max_deg_nodes, key=lambda x: label_rarity[G1_labels[x]])
+
+                node_order.append(next_node)
+                for node in G1.neighbors(next_node):
+                    used_degrees[node] += 1
+
+                nodes_to_add.remove(next_node)
+                label_rarity[G1_labels[next_node]] -= 1
+                V1_unordered.discard(next_node)
+
+    return node_order
+
+
+def _all_argmax(nodes, key_function):
+    best_nodes = []
+    best = -float("inf")
+    for n in nodes:
+        if key_function(n) > best:
+            best = key_function(n)
+            best_nodes = [n]
+            continue
+        if key_function(n) == best:
+            best_nodes.append(n)
+
+    return best_nodes
```

Comment:
Based on @rossbar's comment and @kpetridis24's intuition early on about a two pass approach to finding all the argmax nodes, I set up some quick tests. This function is not optimal in that it is faster to test if new value is less than the best first since that is the most common case. Also, the new value should be stored so it doesn't have to be recomputed.  Those changes cut the time by 30% or so. 

The result is that this function is faster than a two pass solution when there are many ties, and slow when there aren't many ties. They both scale linearly wit the number of nodes. And the difference is 10% slower to 20% faster.  So, readability wins this time!

Let's go back to the two pass arg_max idiom:
```python
            best = min(-label_rarity[G1_labels[x]] for x in V1_unordered)
            rarest_nodes = [n for n in V1_unordered if label_rarity[G1_labels[x]] == best]
...
                best = max(used_degrees[n] for n in nodes_to_add)
                max_ud_nodes = [n for n in nodes_to_add if used_degrees[n] == best]
                best = max(G1.degree[x] for x in max_ud_nodes)
                rarest_nodes = [n for n in max_ud_nodes if used_degree[n] == best]
```